### PR TITLE
Extensions - Merge string extensions and add unit testing

### PR DESCRIPTION
**Pull request by PabstMirror**
Untitled

### CODE REVIEW

File: addons/interact_menu/functions/fnc_renderIcon.sqf [hunk 1]
```diff
@@ -41,7 +41,7 @@ if (_iconFile isEqualTo "") then {
 _text = if (GVAR(UseListMenu)) then {
     format ["<img image='%1' align='left' color='%2'/><t %3>%4</t>", _iconFile, _iconColor, _textSettings, _text]
 } else {
-    format ["<img image='%1' align='center' color='%2'/><br/><t %3 align='center'>%4</t>", _iconFile, _iconColor, _textSettings, "ace_break_line" callExtension _text];
+    format ["<img image='%1' align='center' color='%2'/><br/><t %3 align='center'>%4</t>", _iconFile, _iconColor, _textSettings, ("ace_break_line" callExtension ["breakLine", [_text]]) select 0];
```
Comment (dedmen) [inline addons/interact_menu/functions/fnc_renderIcon.sqf]: isn't breakline also a strings extension?

File: extensions/CMakeLists.txt [hunk 2]
```diff
@@ -147,6 +146,37 @@ if (DEVEL)
     endif()
 endif()
 
+# ------------- GTEST -------------
+# Download and unpack googletest at configure time
+configure_file(CMakeLists.txt.in googletest-download/CMakeLists.txt)
+execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" .
+  RESULT_VARIABLE result
```
[the remark this answers is not in the document]
Comment (dedmen) [reply extensions/CMakeLists.txt]: It's a argument to `execute_process` thus indented. Though I would prefer

```
command( ARG
                 ARG2 )
```
Like is done in L171 add_subdirectory.

File: extensions/strings/ace_parse_imagepath.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * Structured text that usualy has an image:
+ * Example: "<img image='\A3\Ui_f\data\IGUI\Cfg\Actions\open_door_ca.paa' size='2.5' />";
+ *
+ * Returns:
+ * Just the image path or "" if none
+ */
+
+#include "ace_parse_imagepath.hpp"
+
+std::string getImagePathFromStructuredText(const std::string& input) {
```
Comment (dedmen) [inline extensions/strings/ace_parse_imagepath.cpp]: can't we just run a regex here? :D

File: extensions/strings/extension.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+void __stdcall RVExtensionVersion(char* output, int outputSize) {
+    strncpy(output, ACE_FULL_VERSION_STR, outputSize);
+}
+
+void __stdcall RVExtension(char* output, int outputSize, const char* function) {
+    if (!strcmp(function, "version")) {
+        RVExtensionVersion(output, outputSize);
+        return;
+    }
+    strncpy(output, "error", outputSize);
```
Comment (dedmen) [inline extensions/strings/extension.cpp]: Maybe tell the user/script that he/it has to use the args version

File: extensions/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonpas) [reply extensions/CMakeLists.txt]: It's 2 spaces instead of 4.

File: extensions/strings/extension.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+int __stdcall RVExtensionArgs(char* output, int outputSize, const char* function, const char** args, int argsCnt) {
+    if (!strcmp(function, "breakLine")) {
+        if (argsCnt != 1) { return -1002; } // Error: wrong arg count
```
Comment (TheMagnetar) [inline extensions/strings/extension.cpp]: Can we either typedef or enum the errors?

File: extensions/CMakeLists.txt [hunk 6]
```diff
@@ -147,6 +146,37 @@ if (DEVEL)
...
 endif()
 
+# ------------- GTEST -------------
+# Download and unpack googletest at configure time
+configure_file(CMakeLists.txt.in googletest-download/CMakeLists.txt)
+execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" .
+                RESULT_VARIABLE result
+                WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/googletest-download )
+if(result)
+  message(FATAL_ERROR "CMake step for googletest failed: ${result}")
```
Comment (jonpas) [inline extensions/CMakeLists.txt]:
```suggestion
    message(FATAL_ERROR "CMake step for googletest failed: ${result}")
```

File: extensions/CMakeLists.txt [hunk 7]
```diff
@@ -147,6 +146,37 @@ if (DEVEL)
...
+endif()
+execute_process(COMMAND ${CMAKE_COMMAND} --build .
+                RESULT_VARIABLE result
+                WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/googletest-download )
+if(result)
+  message(FATAL_ERROR "Build step for googletest failed: ${result}")
```
Comment (jonpas) [inline extensions/CMakeLists.txt]:
```suggestion
    message(FATAL_ERROR "Build step for googletest failed: ${result}")
```

File: extensions/CMakeLists.txt [hunk 8]
```diff
@@ -147,6 +146,37 @@ if (DEVEL)
...
+# the gtest and gtest_main targets.
+add_subdirectory(${CMAKE_CURRENT_BINARY_DIR}/googletest-src
+                 ${CMAKE_CURRENT_BINARY_DIR}/googletest-build
+                 EXCLUDE_FROM_ALL)
+
+# The gtest/gtest_main targets carry header search path
+# dependencies automatically when using CMake 2.8.11 or
+# later. Otherwise we have to add them here ourselves.
+if (CMAKE_VERSION VERSION_LESS 2.8.11)
+  include_directories("${gtest_SOURCE_DIR}/include")
```
Comment (jonpas) [inline extensions/CMakeLists.txt]:
```suggestion
    include_directories("${gtest_SOURCE_DIR}/include")
```

File: extensions/CMakeLists.txt [hunk 9]
```diff
@@ -147,6 +146,37 @@ if (DEVEL)
...
+# Add googletest directly to our build. This defines
+# the gtest and gtest_main targets.
+add_subdirectory(${CMAKE_CURRENT_BINARY_DIR}/googletest-src
+                 ${CMAKE_CURRENT_BINARY_DIR}/googletest-build
+                 EXCLUDE_FROM_ALL)
+
+# The gtest/gtest_main targets carry header search path
+# dependencies automatically when using CMake 2.8.11 or
+# later. Otherwise we have to add them here ourselves.
+if (CMAKE_VERSION VERSION_LESS 2.8.11)
```
Comment (jonpas) [inline extensions/CMakeLists.txt]: We require CMake 3.0, so this special handling is not necessary.

File: extensions/CMakeLists.txt.in [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# ------------- GTEST -------------
```
Comment (jonpas) [inline extensions/CMakeLists.txt.in]: Put this file into `cmake` folder next to FindDirectX module?

File: extensions/strings/tests/tester.cpp [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        ASSERT_EQ(extReturn, 0);
+    }
+    TEST(breakLine, extension_wrongArgCount) {
+        char output[256];
+        char function[] = "breakLine";
+        int extReturn = RVExtensionArgs(output, 256, function, nullptr, 0);
+        std::cout << "extension_wrongArgCount: " << output << "\n";
+        ASSERT_EQ(extReturn, RETURN_WRONG_ARG_COUNT);
+    }
+
```
Comment (jonpas) [inline extensions/strings/tests/tester.cpp]:
```suggestion
```

File: extensions/strings/tests/tester.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
```
Comment (jonpas) [inline extensions/strings/tests/tester.cpp]:
```suggestion
```

**Comment by PabstMirror** [comment]
PR is adding extra quotes, will try to figure out why

**Comment by jonpas** [comment]
Conflicts and some smaller changes requested, could be merged otherwise.

File: extensions/strings/ace_parse_imagepath.cpp (hunk 3 above)
Comment (PabstMirror) [reply extensions/strings/ace_parse_imagepath.cpp:16]: wouldn't it be nice if arma could do it

Comment (dedmen) [reply extensions/strings/ace_parse_imagepath.cpp:16]: Arma can do it now, 2.06

**Pull request closed without merging**